assets/scripts/modules/Modal.js

```diff
@@ -0,0 +1,195 @@
+
+        // Bindings
+        this.toggle         = this.toggle.bind(this)
+        this.onModalOpen    = this.onModalOpen.bind(this)
+        this.onVisitStart   = this.onVisitStart.bind(this)
+
+        // UI
+        this.$togglers          = document.querySelectorAll(`[data-${this.dataName}-toggler]`)
+        this.$focusTrapTargets  = Array.from(this.el.querySelectorAll(`[data-${this.dataName}-target]`))
+
+        // Focus trap options
+        this.focusTrapOptions = {
+            /**
+             * There is a delay between when the class is applied
+             * and when the element is focusable
+             */
+            checkCanFocusTrap: (trapContainers) => {
+                const results = trapContainers.map((trapContainer) => {
+                    return new Promise((resolve) => {
+                        const interval = setInterval(() => {
+                            if (
+                                getComputedStyle(trapContainer).visibility !==
+                                'hidden'
+                            ) {
+                                resolve()
+                                clearInterval(interval)
+                            }
+                        }, 5)
+                    })
+                })
+
+                // Return a promise that resolves when all the trap containers are able to receive focus
+                return Promise.all(results)
+            },
+
+            onActivate: () => {
+                this.el.classList.add(Modal.CLASS.EL)
+                $html.classList.add(Modal.CLASS.HTML)
+                $html.classList.add('has-'+this.dataName+'-open')
+                this.el.setAttribute('aria-hidden', false)
+                this.isOpen = true
+
+                this.onActivate?.();
+            },
+
+            onPostActivate: () => {
+                this.$togglers.forEach(($toggler) => {
+                    $toggler.setAttribute('aria-expanded', true)
+                })
+            },
+
+            onDeactivate: () => {
+                this.el.classList.remove(Modal.CLASS.EL)
+                $html.classList.remove(Modal.CLASS.HTML)
+                $html.classList.remove('has-'+this.dataName+'-open')
+                this.el.setAttribute('aria-hidden', true)
+                this.isOpen = false
+
+                this.onDeactivate?.();
+            },
+
+            onPostDeactivate: () => {
+                this.$togglers.forEach(($toggler) => {
+                    $toggler.setAttribute('aria-expanded', false)
+                })
+            },
+
+            clickOutsideDeactivates: true,
+        }
+
+        this.isOpen = false
+    }
+
+    /////////////////
+    // Lifecycle
+    /////////////////
+    init() {
+        this.onBeforeInit?.()
+
+        this.focusTrap = createFocusTrap(
+            this.$focusTrapTargets.length > 0 ? this.$focusTrapTargets : [this.el],
+            this.focusTrapOptions
+        )
```

Comment:
```suggestion
        // UI
        this.$togglers          = document.querySelectorAll(`[data-${this.dataName}-toggler]`)
        this.$focusTrapTargets  = Array.from(this.el.querySelectorAll(`[data-${this.dataName}-target]`))

        this.isOpen = false
    }

    getFocusTrapOptions() {
        return {
            /**
             * There is a delay between when the class is applied
             * and when the element is focusable
             */
            checkCanFocusTrap: (trapContainers) => {
                const results = trapContainers.map((trapContainer) => {
                    return new Promise((resolve) => {
                        const interval = setInterval(() => {
                            if (
                                getComputedStyle(trapContainer).visibility !==
                                'hidden'
                            ) {
                                resolve()
                                clearInterval(interval)
                            }
                        }, 5)
                    })
                })

                // Return a promise that resolves when all the trap containers are able to receive focus
                return Promise.all(results)
            },

            onActivate: () => {
                this.el.classList.add(Modal.CLASS.EL)
                $html.classList.add(Modal.CLASS.HTML)
                $html.classList.add('has-'+this.dataName+'-open')
                this.el.setAttribute('aria-hidden', false)
                this.isOpen = true

                this.onActivate?.();
            },

            onPostActivate: () => {
                this.$togglers.forEach(($toggler) => {
                    $toggler.setAttribute('aria-expanded', true)
                })
            },

            onDeactivate: () => {
                this.el.classList.remove(Modal.CLASS.EL)
                $html.classList.remove(Modal.CLASS.HTML)
                $html.classList.remove('has-'+this.dataName+'-open')
                this.el.setAttribute('aria-hidden', true)
                this.isOpen = false

                this.onDeactivate?.();
            },

            onPostDeactivate: () => {
                this.$togglers.forEach(($toggler) => {
                    $toggler.setAttribute('aria-expanded', false)
                })
            },

            clickOutsideDeactivates: true,
        }
    }

    /////////////////
    // Lifecycle
    /////////////////
    init() {
        this.onBeforeInit?.()

        this.focusTrap = createFocusTrap(
            this.$focusTrapTargets.length > 0 ? this.$focusTrapTargets : [this.el],
            this.getFocusTrapOptions()
        )
```

`this.focusTrapOptions` should be moved to a separate method, something like `this.getFocusTrapOptions` which would be easier to override/extend in a sub-class.